benchmarks/kslookup/hsearch.c

```diff
@@ -0,0 +1,309 @@
+{
+	ENTRY * ep;
+	struct timeval start;
+	struct timeval end;
+
+	gettimeofday (&start, 0);
+//MEASURE START
+
+	//create
+	hcreate (k);
+	//insert
+	for (int i = 0;i < n;++i)
+	{
+		ep = hsearch (data[i], ENTER);
+		if (ep == NULL)
+		{
+			fprintf (stderr, "ENTER failed\n");
+			return -1;
+		}
+	}
+//MEASURE END
+	gettimeofday (&end, 0);
+	times[r] = (int) (end.tv_sec - start.tv_sec) * 1000000 +
+						(end.tv_usec - start.tv_usec);
+
+	//cleanup
+	hdestroy ();
+	return 1;
+}
+
+/* The actual benchmark procedure for the search, executed for each
+ * setting.
+ */
+int search (int n, int k, int r, ENTRY * data, int * times)
+{
+	ENTRY * ep;
+	ENTRY e;
+
+	struct timeval start;
+	struct timeval end;
+	//this array saves all the times to search each key
+	int keys_searched_for[n];
+
+	for (int i = 0;i < n;++i) keys_searched_for[i] = -1;
+
+	//prepare hash map
+	hcreate (k);
+	for (int i = 0;i < n;++i)
+	{
+		ep = hsearch (data[i], ENTER);
+		if (ep == NULL)
+		{
+			fprintf (stderr, "ENTER failed\n");
+			return -1;
+		}
+	}
+
+	// search for each key randomly and save the time
+	for(int i = 0;i < n;++i)
+	{
+		int search_for = searchNext (keys_searched_for, n);
+		char * lookfor = data[search_for].key;
+
+		e.key = lookfor;
+
+		gettimeofday (&start, 0);
+//MEASURE START
+
+		ep = hsearch(e, FIND);
+//MEASURE END
+		gettimeofday (&end, 0);
+		keys_searched_for[search_for] = (int) (end.tv_sec - start.tv_sec) * 1000000 +
+							(end.tv_usec - start.tv_usec);
+
+		if (ep == NULL)
+		{
+			fprintf (stderr, "not found while search\n");
+			return -1;
+		}
+		Key * validate = ep->data;
+		if (strcmp (keyName(validate), lookfor) != 0)
+		{
+			fprintf (stderr, "found wrong Key while search\n");
+			return -1;
+		}
+		if (strcmp (keyValue(validate), GENDATA_KEY_VALUE) != 0)
+		{
+			fprintf (stderr, "wrong Key value while search\n");
+			return -1;
+		}
+	}
+
+	// take the maximum for this run
+	times[r] = max (keys_searched_for, n);
+
+	//cleanup
+	hdestroy ();
+	return 1;
+}
+
+KeySet * readKeySet (int size, int version)
+{
+	KeySet * out = ksNew(size, KS_END);
+	//plugin open
+	KeySet * modules = ksNew(0, KS_END);
+	elektraModulesInit(modules, 0);
+	KeySet *conf = ksNew (0, KS_END);
+	Plugin * plugin = elektraPluginOpen(EXPORT_PLUGIN, modules, conf, 0);
+	if(!plugin)
+	{
+		fprintf (stderr, "dump plugin could not be opened\n");
+		exit (EXIT_FAILURE);
+	}
+
+	char filename_in [BUFFER_FILENAME];
+	sprintf (&filename_in[0], "%i_%i.edf",size,version);
+
+	Key * pkey = keyNew(0);
+	keySetString (pkey, &filename_in[0]);
+
+	plugin->kdbGet (plugin, out, pkey);
+
+	const Key * ekey = keyGetMeta (pkey,"error/description");
+	if (ekey != 0)
+	{
+		fprintf (stderr, "%s\n", (char *) keyString(ekey));
+		exit (EXIT_FAILURE);
+	}
+
+	keyDel (pkey);
+
+	//plugin close
+	elektraPluginClose (plugin, 0);
+	elektraModulesClose(modules, 0);
+	ksDel (modules);
+
+	return out;
+}
+
+/* generates the ENTRYs for the hash map
+ */
+ENTRY * prepareData (KeySet * ks)
```

Comment:
Where is ENTRY defined? ENTRY seems to be a too general name. And why is it a macro?